### PR TITLE
composite-checkout: Refactor payment method creation to remove Container

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

In #38727 we will be adding a new payment method for full credits payments. 

In order to implement this, some refactoring had to be done, which makes the other PR rather unfortunately large. This PR is the unrelated refactoring only.

Previously, the payment method objects were being created in the `CompositeCheckoutContainer`, which then passed them down to the `CompositeCheckout` component in Calypso to in turn pass them to the `CheckoutProvider` where they are used. However, we only learn about the number of available credits when we download the shopping cart with the `useShoppingCart` hook, and that happens in `CheckoutProvider`. Therefore, this PR moves the payment method creation down to that component.

With that code moved (I moved most of it into a separate file to make it easier to grok since it is getting rather large and there's a lot of caching involved), `CompositeCheckoutContainer` was starting to be rather small, and it seemed to me that there's really no need for it any longer (it was originally present because `CompositeCheckout` was not in the Calypso package, but now it is). Therefore, I moved the remainder of its logic into `CompositeCheckout` and removed the now-unused class. 

This caused some of the tests to fail (for reasons I still don't fully understand) and so I also refactored the tests to make them work again. One of the things I had to do was to allow passing in an array of allowed payment method strings to `CompositeCheckout` to filter the payment methods that would be created. This will be useful later when we need to also use the `allowed_payment_methods` array from the cart endpoint.

Because the payment methods themselves are created before the checkout form is rendered, and two of the payment method rely on async data (credits and existing payment methods), we also needed to refactor when the payment methods get created and how they are cached. It made more sense to me to wait for the payment methods to be fully created before showing the form. Therefore, this PR also treats loading stored cards as a loading state, and moves the payment method caching to the inside of the `CompositeCheckout` component. This allows us to return an empty array of payment methods while things are still loading.

#### Testing instructions

- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`.
- Sandbox the store.
- Add a plan to your cart.
- Complete checkout with a card (new or saved) and verify that everything works as expected.

### CODE REVIEW

**Comment by nbloomf** [comment]
I'm still testing this, but just noticed one weird behavior I don't remember seeing before. When checkout was first loading I thought it got stuck in a 'loading' state for a while, but it appears the component just didn't rerender until I moused over the page. (This may be normal and I've just never noticed it before.)

**Comment by nbloomf** [comment]
Also ran into a weird situation that is probably not related to this PR - the steps went like this.

1. Enter checkout with a domain
2. Make a mistake in the phone number field
3. Continue to step 3 and then pay
4. Get an error like "Enter a valid country code followed by a dot (for example +1.6285550199)."
5. Edit step 2 and fix the error, continue to step 3 again and then pay.
6. (This is the confusing part) Get the same error message a second time
7. While trying to figure out the mistake in my phone number, proceed to the thank you page

I suspect there's a caching bug where the error state isn't getting cleared at the right time.

File: packages/composite-checkout-wpcom/src/mock/cart-endpoint.ts [hunk 1]
```diff
@@ -7,26 +7,15 @@ import { RequestCart, ResponseCart } from '../types';
 
 /**
  * A fake WPCOM shopping cart endpoint.
- *
- * This is awful.
```
Comment (nbloomf) [inline packages/composite-checkout-wpcom/src/mock/cart-endpoint.ts]: 😂

**Review by nbloomf — APPROVED** [review]
Looks good and behaves as expected.

**Review by fditrapani — APPROVED** [review]
I tested this and it works as expected.

**Comment by sirbrillig** [comment]
> I suspect there's a caching bug where the error state isn't getting cleared at the right time.

I think I've seen something similar before. 

I don't think this is related to this PR, so I've added a card to the project to check it out.

**Pull request merged**